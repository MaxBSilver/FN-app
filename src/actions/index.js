export const isLoading = bool => ({
  type: "IS_LOADING",
  isLoading: bool
});

export const hasErrored = message => ({
  type: "HAS_ERRORED",
  message
});

export const setNews = news => ({
  type: "SET_NEWS",
  news
});

export const setStore = store => ({
  type: "SET_STORE",
  store
});
