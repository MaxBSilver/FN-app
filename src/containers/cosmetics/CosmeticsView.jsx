import React, { Component } from "react";
import { connect } from "react-redux";
import { getAllCosmeticsThunk } from "../../api/cosmetics";
import CosmeticItemContainer from "../../components/cosmeticItemContainer/CosmeticItemContainer";
import Loading from "../../components/loading/Loading";
import PropTypes from "prop-types";
export class CosmeticsView extends Component {
  constructor(props) {
    super(props);
    this.state = {
      filter: "all",
      displayItems: []
    };
  }
  async componentDidMount() {
    if (this.props.cosmeticItems.length === 0)
      await this.props.getAllCosmeticsThunk();
    this.setState({ displayItems: this.props.cosmeticItems });
  }

  selectDisplay = (filter) => {
    const { cosmeticItems } = this.props;
    const displayItems = cosmeticItems.filter(
      cosmetic => cosmetic.item.type === filter
    );
    this.setState({ displayItems, filter });
  };

  handleClick = e => {
    const { name } = e.target;
    this.selectDisplay(name);
  };

  render() {
    return (
      <section className="cosmetics-container">
        {!this.props.isLoading && (
          <React.Fragment>
            <h2>Cosmetics</h2>
            <div className="control-container">
              <button
                className="gliders-btn"
                name="glider"
                onClick={this.handleClick}
              >
                Gliders
              </button>
              <button
                className="pickaxes-btn"
                name="pickaxe"
                onClick={this.handleClick}
              >
                Pickaxes
              </button>
              <button
                className="outfits-btn"
                name="outfit"
                onClick={this.handleClick}
              >
                Outfits
              </button>
            </div>
            <section className="cosmetic-item-container">
              <CosmeticItemContainer displayItems={this.state.displayItems} />
            </section>
          </React.Fragment>
        )}
        {this.props.isLoading && <Loading />}
      </section>
    );
  }
}

export const mapStateToProps = state => ({
  cosmeticItems: state.cosmeticItems,
  isLoading: state.isLoading,
  error: state.error
});

export const mapDispatchToProps = dispatch => ({
  getAllCosmeticsThunk: () => dispatch(getAllCosmeticsThunk())
});
CosmeticsView.propTypes = {
  cosmeticItems: PropTypes.array,
  error: PropTypes.string,
  getAllCosmeticsThunk: PropTypes.func,
  isLoading: PropTypes.bool
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CosmeticsView);
