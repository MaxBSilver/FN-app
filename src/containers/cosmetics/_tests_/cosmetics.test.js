import React from "react";
import { CosmeticsView, mapStateToProps } from "../CosmeticsView";
import * as action from "../../../actions/index";
import { shallow } from "enzyme";

describe("CosmeticsView", () => {
  describe("component", () => {
    let wrapper;
    let mockGetAllCosmeticsThunk;
    beforeEach(() => {
      mockGetAllCosmeticsThunk = jest.fn();
      const mockCosmeticItems = [];
      wrapper = shallow(
        <CosmeticsView
          getAllCosmeticsThunk={mockGetAllCosmeticsThunk}
          cosmeticItems={mockCosmeticItems}
        />
      );
    });
    it("should match snapshot", () => {
      expect(wrapper).toMatchSnapshot();
    });
    it("should invoke getAllCosmeticsThunk when component mounts", async () => {
      await wrapper.instance().componentDidMount();
      expect(mockGetAllCosmeticsThunk).toHaveBeenCalled();
    });
    it("should have a default state", () => {
      expect(wrapper.instance().state).toEqual({
        filter: "all",
        displayItems: []
      });
    });
    describe("handle click", () => {
      it("should invoke select display on click", () => {
        wrapper.instance().selectDisplay = jest.fn();
        const event = { target: "max" };
        wrapper.instance().handleClick(event);
        expect(wrapper.instance().selectDisplay).toHaveBeenCalled();
      });
      describe("selectDisplay", () => {
        const mockCosmeticItems = [{ item: { type: "pickaxe" } }, { item: { type: "glider" } }];

        wrapper = shallow(
          <CosmeticsView
            getAllCosmeticsThunk={mockGetAllCosmeticsThunk}
            cosmeticItems={mockCosmeticItems}
          />
        );
        it("should setState with an item with a type equivalent to the filter", () => {
          wrapper.instance().selectDisplay('pickaxe')
          expect(wrapper.instance().state).toEqual({displayItems: [{ item: { type: "pickaxe" } }], filter: 'pickaxe'})        });
      });
    });
  });

  describe("mapStateToProps", () => {
    it("should return an object with the cosmeticItems array", () => {
      const mockState = {
        cosmeticItems: [{ title: "New Skins" }, { title: "New Items" }]
      };
      const expected = {
        cosmeticItems: [{ title: "New Skins" }, { title: "New Items" }]
      };
      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected);
    });
    it("should return an boolean", () => {
      const mockState = {
        isLoading: false
      };
      const expected = {
        isLoading: false
      };
      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected);
    });
    it("should return an error message", () => {
      const mockState = {
        error: "Has Errored!"
      };
      const expected = {
        error: "Has Errored!"
      };
      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected);
    });
  });
});
