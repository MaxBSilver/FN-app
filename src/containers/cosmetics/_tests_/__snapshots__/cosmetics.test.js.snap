// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CosmeticsView component should match snapshot 1`] = `ShallowWrapper {}`;

exports[`CosmeticsView component should match snapshot 2`] = `ShallowWrapper {}`;

exports[`CosmeticsView should match snapshot 1`] = `ShallowWrapper {}`;
